### PR TITLE
Fix RTM backend

### PR DESCRIPTION
**Pull request by ostefano**
This is not ready to be merged, but rather an ask for feedback:

- I think that during codebase transitions it would be nice to allow 'forcing' of a specific backend architecture (see `BACKEND_ARCH`)
- Some of the RTM methods are still without a cursor or a limit. I have replaced one with a 'limited' version, but I was wondering if all methods should use a cursor?
- Accessing `username_to_userid` is super expensive on workspaces with > 30K users, so I have tried to optimize a couple of cases
- Restarting the bot using the RTM backend takes more than 20 seconds, as if a timeout needs to occur
- I see a strange behaviour where the bot is restarted, and some plugins (specifically the one related to handling `TextCmds`) are initialised twice or even more times, resulting in bot issued messages being delivered multiple times (see logs)

```2021-02-01 12:57:36,855 INFO     errbot.core               Notifying connection to all the plugins...
2021-02-01 12:57:36,855 INFO     errbot.plugins.ChatRoom   Connecting bot chatrooms
2021-02-01 12:57:36,855 INFO     errbot.core               Plugin activation done.
2021-02-01 12:57:36,855 INFO     errbot.core               Activate internal commands
2021-02-01 12:57:36,855 INFO     errbot.plugin_manager     Activate bot plugins...
2021-02-01 12:57:36,855 INFO     errbot.plugin_manager     Activate plugin: Webserver.
2021-02-01 12:57:36,856 INFO     errbot.plugins.Webserver  Webserver is not configured. Forbid activation
2021-02-01 12:57:36,856 INFO     errbot.core_plugins.wsview Checking Webserver for webhooks
2021-02-01 12:57:36,856 INFO     errbot.core_plugins.wsview Webhook routing echo
2021-02-01 12:57:36,856 INFO     errbot.plugin_manager     Activate plugin: TextCmds.
2021-02-01 12:57:36,856 INFO     errbot.core_plugins.wsview Checking TextModeCmds for webhooks
2021-02-01 12:57:36,857 INFO     errbot.core               Notifying connection to all the plugins...
2021-02-01 12:57:36,857 INFO     errbot.plugins.ChatRoom   Connecting bot chatrooms
2021-02-01 12:57:36,857 INFO     errbot.core               Plugin activation done.
2021-02-01 12:57:36,857 INFO     errbot.core               Activate internal commands
2021-02-01 12:57:36,857 INFO     errbot.plugin_manager     Activate bot plugins...
2021-02-01 12:57:36,857 INFO     errbot.plugin_manager     Activate plugin: Webserver.
2021-02-01 12:57:36,857 INFO     errbot.plugins.Webserver  Webserver is not configured. Forbid activation
2021-02-01 12:57:36,858 INFO     errbot.core_plugins.wsview Checking Webserver for webhooks
2021-02-01 12:57:36,858 INFO     errbot.core_plugins.wsview Webhook routing echo
2021-02-01 12:57:36,858 INFO     errbot.plugin_manager     Activate plugin: TextCmds.
2021-02-01 12:57:36,858 INFO     errbot.core_plugins.wsview Checking TextModeCmds for webhooks
2021-02-01 12:57:36,858 INFO     errbot.core               Notifying connection to all the plugins...
2021-02-01 12:57:36,858 INFO     errbot.plugins.ChatRoom   Connecting bot chatrooms
2021-02-01 12:57:36,859 INFO     errbot.core               Plugin activation done.
2021-02-01 12:58:26,043 INFO     errbot.plugins.ACLs       Matching ACL {} against username @testuser for command Help:about.
2021-02-01 12:58:26,044 INFO     errbot.core               Processing command "about" with parameters "" from #yet-another-private/testuser
2021-02-01 12:58:26,679 INFO     errbot.plugins.ACLs       Matching ACL {} against username @testuser for command Help:about.
2021-02-01 12:58:26,680 INFO     errbot.core               Processing command "about" with parameters "" from #yet-another-private/testuser
2021-02-01 12:58:27,327 INFO     errbot.plugins.ACLs       Matching ACL {} against username @testuser for command Help:about.
2021-02-01 12:58:27,328 INFO     errbot.core               Processing command "about" with parameters "" from #yet-another-private/testuser```

### CODE REVIEW

**Comment by nzlosh** [comment]
Thanks for the PR.  Optimisations to improve performance on large scale deployments are most welcome.

> ** think that during codebase transitions it would be nice to allow 'forcing' of a specific backend architecture (see BACKEND_ARCH)**

A legacy bot token can only be used with RTM.  Current bot tokens can only be used with the Event API.  The bot makes the distinction between Request URL or Socket Mode by the presence of the `app_token`.  The point of this work is to make transitions as transparent to the end user as possible.  This gives enough control over which architecture to use so I'm not clear on what a flag will gain.  What are the thoughts behind a flag to force an architecture model?

> Restarting the bot using the RTM backend takes more than 20 seconds, as if a timeout needs to occur

I am not seeing those delays with my setup (~3 seconds for a restart).  Could the delay you're seeing be related to the size of the Slack workspace the bot is connecting to?
```
2021-02-01 14:35:55,139 DEBUG    errbot.core               *** frm = @nzlosh
2021-02-01 14:35:55,140 DEBUG    errbot.core               *** username = @nzlosh
2021-02-01 14:35:55,140 DEBUG    errbot.core               *** text = .b restart
2021-02-01 14:35:55,140 DEBUG    errbot.plugins.ACLs       Check restart for ACLs.
2021-02-01 14:35:55,140 INFO     errbot.plugins.ACLs       Matching ACL {} against username @nzlosh for command Health:restart.
...
2021-02-01 14:35:58,699 DEBUG    errbot.plugin_manager     Activate flow plugins ...
2021-02-01 14:35:58,699 INFO     errbot.core               Notifying connection to all the plugins...
2021-02-01 14:35:58,700 DEBUG    errbot.core               Trigger callback_connect on Flows.
2021-02-01 14:35:58,700 DEBUG    errbot.core               Trigger callback_connect on Webserver.
2021-02-01 14:35:58,700 DEBUG    errbot.core               Trigger callback_connect on VersionChecker.
2021-02-01 14:35:58,700 DEBUG    errbot.core               Trigger callback_connect on Health.
2021-02-01 14:35:58,702 DEBUG    errbot.core               Trigger callback_connect on CommandNotFoundFilter.
2021-02-01 14:35:58,703 DEBUG    errbot.core               Trigger callback_connect on Plugins.
2021-02-01 14:35:58,705 DEBUG    errbot.core               Trigger callback_connect on Backup.
2021-02-01 14:35:58,707 DEBUG    errbot.core               Trigger callback_connect on Help.
2021-02-01 14:35:58,708 DEBUG    errbot.core               Trigger callback_connect on ACLS.
2021-02-01 14:35:58,709 DEBUG    errbot.core               Trigger callback_connect on ChatRoom.
2021-02-01 14:35:58,711 INFO     errbot.plugins.ChatRoom   Connecting bot chatrooms
2021-02-01 14:35:58,712 DEBUG    errbot.core               Trigger callback_connect on Utils.
2021-02-01 14:35:58,713 DEBUG    errbot.core               Trigger callback_connect on Example.
2021-02-01 14:35:58,713 INFO     errbot.core               Plugin activation done.
```

> I see a strange behaviour where the bot is restarted, and some plugins (specifically the one related to handling TextCmds) are initialised twice or even more times, resulting in bot issued messages being delivered multiple times (see logs)

I've also seen strange behaviour with the RTM client. I've been looking into the issue but haven't completely isolated the source of the problem.

**Comment by nzlosh** [comment]
I've created an issue for the duplicate messages.  https://github.com/nzlosh/err-backend-slackv3/issues/12

**Comment by ostefano** [comment]
> A legacy bot token can only be used with RTM. Current bot tokens can only be used with the Event API. The bot makes the distinction between Request URL or Socket Mode by the presence of the app_token. The point of this work is to make transitions as transparent to the end user as possible. This gives enough control over which architecture to use so I'm not clear on what a flag will gain. What are the thoughts behind a flag to force an architecture model?

For example, in my case I have a token that works with SlackRTM but does not work with the Event API because I don't have a signing key. Forcing RTM is the only way for me to use this backend.

> I am not seeing those delays with my setup (~3 seconds for a restart). Could the delay you're seeing be related to the size of the Slack workspace the bot is connecting to?

Won't be the first case, but it's still weird to see something like this at disconnection time.

> I've also seen strange behaviour with the RTM client. I've been looking into the issue but haven't completely isolated the source of the problem.

What is weird is that it seems to affect only the core plugins

**Comment by ostefano** [comment]
> I am not seeing those delays with my setup (~3 seconds for a restart). Could the delay you're seeing be related to the size of the Slack workspace the bot is connecting to?

@nzlosh wait, my bad here. I am restarting the bot using a systemd unit file, which is sending a SIGINT as a KillSignal. Might this be related to other issues of the core API registering for specific signals?

**Comment by duhow** [comment]
- `cursor` should only used when calling `.list` API calls, I'm not sure if the slack-api can perform a `query_all` argument to do so. Otherwise will need to create functions in the backend itself to call these.
- Regarding `username_to_userid` and other related API calls , I really consider we should be using the Errbot Storage system to cache data. Some may be more sustained to have changes, but still would be good to have something to start.

**Comment by ostefano** [comment]
@nzlosh 

> cursor should only used when calling .list API calls, I'm not sure if the slack-api can perform a query_all argument to do so. Otherwise will need to create functions in the backend itself to call these.

Perfect, then I should change my path for `username_to_userid` to use users_list with `cursor` rather than a limit.

> regarding username_to_userid and other related API calls , I really consider we should be using the Errbot Storage system to cache data. Some may be more sustained to have changes, but still would be good to have something to start.

Even a LRU in-process cache would do. Why would `Errbot Storage system` be more advisable?

**Comment by nzlosh** [comment]
I'm interested in knowing a bit more about your slack token for SlackRTM.  Would you mind running the bot in debug mode and sharing the "Verifying authentication token" call (obfuscate any sensitive data)?  I'm specifically interested in the `x-oauth-scopes` field in the response.

> @nzlosh wait, my bad here. I am restarting the bot using a systemd unit file, which is sending a SIGINT as a KillSignal. Might this be related to other issues of the core API registering for specific signals?
>
> Maybe related to this #7 ?

Yes, I'd say so.  I'm currently working on a patch, but it's still not sufficiently stable to create a PR.

**Comment by ostefano** [comment]
@nzlosh there you go
```
2021-02-01 14:19:03,967 DEBUG    slack_sdk.web.base_client Sending a request - url: https://www.slack.com/api/auth.test, query_params: {}, body_params: {}, files: {}, json_body: {}, headers: {'Content-Type': 'application/json;charset=utf-8', 'Authorization': '(redacted)', 'User-Agent': 'Python/3.6.10 slackclient/3.2.1 Linux/4.15.0-1080-gcp'}
2021-02-01 14:19:04,175 DEBUG    slack_sdk.web.slack_response Received the following response - status: 200, headers: {'date': 'Mon, 01 Feb 2021 14:19:04 GMT', 'server': 'Apache', 'x-xss-protection': '0', 'pragma': 'no-cache', 'cache-control': 'private, no-cache, no-store, must-revalidate', 'access-control-allow-origin': '*', 'x-slack-req-id': '(redacted)', 'strict-transport-security': 'max-age=31536000; includeSubDomains; preload', 'access-control-expose-headers': 'x-slack-req-id, retry-after', 'x-slack-backend': 'r', 'x-oauth-scopes': 'identify,read,post,client,apps', 'referrer-policy': 'no-referrer', 'x-content-type-options': 'nosniff', 'expires': 'Mon, 26 Jul 1997 05:00:00 GMT', 'vary': 'Accept-Encoding', 'access-control-allow-headers': 'slack-route, x-slack-version-ts, x-b3-traceid, x-b3-spanid, x-b3-parentspanid, x-b3-sampled, x-b3-flags', 'content-type': 'application/json; charset=utf-8', 'x-envoy-upstream-service-time': '12', 'x-backend': 'main_normal main_canary_with_overflow main_control_with_overflow', 'x-server': 'slack-www-hhvm-main-iad-6iup', 'x-via': 'envoy-www-iad-1s2s, haproxy-edge-iad-l1sp', 'x-slack-shared-secret-outcome': 'shared-secret', 'via': 'envoy-www-iad-1s2s', 'connection': 'close', 'transfer-encoding': 'chunked'}, body: {'ok': True, 'url': 'https://(redacted).slack.com/', 'team': '(redacted)', 'user': '(redacted)', 'team_id': '(redacted)', 'user_id': '(redacted)', 'bot_id': '(redacted)', 'enterprise_id': '(redacted)'}
2021-02-01 14:19:04,175 DEBUG    errbot.backends.slackv3   Auth response: {'ok': True, 'url': '(redacted)', 'team': 'VMware', 'user': '(redacted)', 'team_id': '(redacted)', 'user_id': '(redacted)', 'bot_id': '(redacted)', 'enterprise_id': '(redacted)'}
2021-02-01 14:19:04,175 INFO     errbot.backends.slackv3   Token accepted
```

**Comment by nzlosh** [comment]
Thanks @ostefano.  What does your slack token start with?  From what I've read, it should be `xoxb-` or `xoxp-`

**Comment by ostefano** [comment]
@nzlosh  that would be `xoxb-`

**Comment by nzlosh** [comment]
Thanks for the information @ostefano.  I've created a [patch](https://github.com/nzlosh/err-backend-slackv3/pull/14) that should correctly identify your token and use the RTMClient.

**Comment by ostefano** [comment]
@nzlosh @duhow I just rebased against your latest commits. Pull requests now have the following fixes:
* fix ACL check against username and not user ids
* fix fetching and storing user profile
* fix using cursors and limit on functions requiring so (turns out limit is mandatory when using a cursor)
* fix _handle_message referencing non-existing attribute .name
* referencing user_id right away when calling send_message rather than asking the API to convert it

File: slackv3.py [hunk 1]
```diff
@@ -475,7 +477,7 @@ def serve_once(self):
         self.update_alternate_prefixes()
 
         # detect api type based on auth_test response (bot:basic is a legacy token scope for RTM)
-        if "bot:basic" in self.auth.headers["x-oauth-scopes"]:
+        if self.backend_arch == "rtm" or "bot:basic" in self.auth.headers["x-oauth-scopes"]:
```
Comment (nzlosh) [inline slackv3.py:480]: Can you revert the `backend_arch` flag since this is handled by this PR https://github.com/nzlosh/err-backend-slackv3/pull/14

Comment (ostefano) [reply slackv3.py:480]: Sure. Since it's not merged yet, I will cherry-pick your fix and merge it here.

Comment (ostefano) [reply slackv3.py]: Done. @nzlosh note that your fix didn't work as-is. I had to add 'apps'.

**Comment by nzlosh** [comment]
Is it possible to break this PR into smaller more specific ones?  There are different issues being addressed in this PR where @duhow would be better positioned to comment on and merge.  Having a PR per issue will help reviewing the code and limit discussions to a specific context.  It would help greatly with merging to main.

**Comment by ostefano** [comment]
@nzlosh yes, I can do that

**Comment by nzlosh** [comment]
@ostefano thanks for agreeing to that.  I've merge #14 so things will be easier for you to fix.

**Comment by ostefano** [comment]
@nzlosh no problem, but the merge of https://github.com/nzlosh/err-backend-slackv3/pull/14 seems to not have included my fix (the 'apps' addition). Do you want that to be yet another pull request?

**Comment by nzlosh** [comment]
@ostefano no, in my haste, I'd forgot to push my change.  I'll add it now.

**Comment by nzlosh** [comment]
@ostefano main branch includes apps, you should be clear to go ahead with your PRs.

**Comment by ostefano** [comment]
Cool, I will rebase and start creating multiple PRs. I will likely be done by tomorrow morning.

**Comment by ostefano** [comment]
@nzlosh created three pull requests:

https://github.com/nzlosh/err-backend-slackv3/pull/17

https://github.com/nzlosh/err-backend-slackv3/pull/18

https://github.com/nzlosh/err-backend-slackv3/pull/19

and converted this PR to draft.

**Pull request closed without merging**